x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -932,7 +933,9 @@ private RangeAvailableHandler readerWithOffset(RangeAvailableHandler reader, Cac
             if (Assertions.ENABLED) {
                 return (channel, channelPos, relativePos, len) -> {
                     assert assertValidRegionAndLength(fileRegion, channelPos, len);
-                    return adjustedReader.onRangeAvailable(channel, channelPos, relativePos, len);
+                    final int bytesRead = adjustedReader.onRangeAvailable(channel, channelPos, relativePos, len);
+                    assert regionOwners.get(fileRegion.io) == fileRegion;
```

Comment:
Maybe assert before calling the reader? Also, can we add a message with the fileRegion (in both assertions)?

Reply:
Similarly, it's already. asserted before the call.

> can we add a message with the fileRegion (in both assertions)?

Good call. I'll do that.

Reply:
Oh right, thanks :+1:

Reply:
pushed ef0d225c0a6ede4c7aea80526ab3a494ea5e11a3